inox2d/src/render.rs

```diff
@@ -1,378 +1,386 @@
+		let nodes = &puppet.nodes;
+		let comps = &mut puppet.node_comps;
+
 		let mut vertex_buffers = VertexBuffers::default();
-		let mut root_drawables_zsorted: Vec<InoxNodeUuid> = Vec::new();
-		let mut node_render_ctxs = HashMap::new();
-
-		for uuid in nodes.all_node_ids() {
-			let node = nodes.get_node(uuid).unwrap();
-
-			node_render_ctxs.insert(
-				uuid,
-				NodeRenderCtx {
-					trans: Mat4::default(),
-					trans_offset: node.trans_offset,
-					kind: match node.data {
-						InoxData::Part(ref part) => {
-							let (index_offset, vert_offset) = vertex_buffers.push(&part.mesh);
-							RenderCtxKind::Part(PartRenderCtx {
+
+		let mut root_drawables_count: usize = 0;
+		for node in nodes.iter() {
+			let drawable_kind = DrawableKind::new(node.uuid, comps);
+			if let Some(drawable_kind) = drawable_kind {
+				root_drawables_count += 1;
+
+				match drawable_kind {
+					DrawableKind::TexturedMesh(components) => {
+						let (index_offset, vert_offset) = vertex_buffers.push(&components.data.mesh);
+						let (index_len, vert_len) =
+							(components.data.mesh.indices.len(), components.data.mesh.vertices.len());
+
+						comps.add(
+							node.uuid,
+							TexturedMeshRenderCtx {
 								index_offset,
 								vert_offset,
-								index_len: part.mesh.indices.len(),
-								vert_len: part.mesh.vertices.len(),
+								index_len,
+								vert_len,
+							},
+						);
+						comps.add(node.uuid, DeformStack::new(vert_len));
+					}
+					DrawableKind::Composite { .. } => {
+						// exclude non-drawable children
+						let children_list: Vec<InoxNodeUuid> = nodes
+							.get_children(node.uuid)
+							.filter_map(|n| {
+								if DrawableKind::new(n.uuid, comps).is_some() {
+									Some(n.uuid)
+								} else {
+									None
+								}
 							})
-						}
+							.collect();
+
+						// composite children are excluded from root_drawables_zsorted
+						root_drawables_count -= children_list.len();
+
+						comps.add(
+							node.uuid,
+							CompositeRenderCtx {
+								// sort later, before render
+								zsorted_children_list: children_list,
+							},
+						);
+					}
+				};
+			}
+		}
 
-						InoxData::Composite(_) => RenderCtxKind::Composite(nodes.zsorted_children(uuid)),
+		let mut root_drawables_zsorted = Vec::new();
+		// similarly, populate later, before render
+		root_drawables_zsorted.resize(root_drawables_count, InoxNodeUuid(0));
 
-						_ => RenderCtxKind::Node,
-					},
-				},
-			);
+		Self {
+			vertex_buffers,
+			root_drawables_zsorted,
 		}
+	}
 
-		for uuid in nodes.zsorted_root() {
-			let node = nodes.get_node(uuid).unwrap();
+	/// Reset all `DeformStack`.
+	pub(crate) fn reset(&mut self, nodes: &InoxNodeTree, comps: &mut World) {
+		for node in nodes.iter() {
+			if let Some(DrawableKind::TexturedMesh(..)) = DrawableKind::new(node.uuid, comps) {
+				let deform_stack = comps
+					.get_mut::<DeformStack>(node.uuid)
+					.expect("A TexturedMesh must have an associated DeformStack.");
```

Comment:
I see a few of these `DeformStack` component fetches. The best thing to do here in case it doesn't exist is to do nothing. This is again a case where you'd log an error at initialization time if the puppet is wrong.

Reply:
My initial response would be: If rendering is initialized, then a `TexturedMesh` MUST have a `DeformStack`, because without a deform (outputted by the respective stack), the shader just don't work. This is not a sign of wrong puppet. This is a sign of wrong rendering initialization logic, which is our problem. So probably better to panic then to continue silently.

My second response: Oh, you give me the insight that we can NOT create a `DeformStack` for a `TexturedMesh` that is not deformed by any parameter (or, the parameter system is not initialized at all), and leave a default buffer region with zeroes for the GPU. In this case yeah we can just skip `None`.